src/devices/src/virtio/net/device.rs

```diff
@@ -405,29 +413,40 @@ impl Net {
 
     // Tries to detour the frame to MMDS and if MMDS doesn't accept it, sends it on the host TAP.
     //
-    // `frame_buf` should contain the frame bytes in a slice of exact length.
     // Returns whether MMDS consumed the frame.
     fn write_to_mmds_or_tap(
         mmds_ns: Option<&mut MmdsNetworkStack>,
         rate_limiter: &mut RateLimiter,
-        frame_buf: &[u8],
+        headers: &mut [u8],
+        frame_iovec: &IoVecBuffer,
         tap: &mut Tap,
         guest_mac: Option<MacAddr>,
     ) -> Result<bool> {
-        let checked_frame = |frame_buf| {
-            frame_bytes_from_buf(frame_buf).map_err(|e| {
-                error!("VNET header missing in the TX frame.");
-                METRICS.net.tx_malformed_frames.inc();
-                e
-            })
-        };
+        // Read the frame headers from the IoVecBuffer. This will return None
+        // if the frame_iovec is empty.
+        let header_len = frame_iovec.read_at(headers, 0).ok_or_else(|| {
```

Comment:
* Is there a benefit to using [`ok_or_else`](https://docs.rs/futures-async-combinators/0.3.1/futures_async_combinators/future/fn.or_else.html) over [`map_err`](https://docs.rs/futures-async-combinators/0.3.1/futures_async_combinators/future/fn.map_err.html)?
* The previous code implied that a missing header was the only reason for an Err to be thrown here, are there any possibilities of other errors being thrown, and if so should we be wrapping it rather? (as far as I can tell we are dropping it here)

Reply:
`ok_or_else` is used here because `read_at` returns an `Option` not a `Result`.  The resulting `Option` will be `None` if we cannot read enough bytes, i.e. there is not a complete header in the frame. In that case we will return `Error::VnetHeaderMissing` (note the `?` operator in the end of the expression).

If not, we will continue, having read `header_len` bytes inside the `headers` buffer.